### ISSUE TITLE
Idea for AllowedHeaders

### ISSUE THREAD

**Issue by xogeny** [rs/cors#4]
The specification doesn't allow the "AllowedHeaders" to be a wildcard.  But what do you think of the idea of checking to see if AllowedHeaders is just `[]string{"*"}` and, if so, simply echo back whatever `Access-Control-Request-Headers` happens to be.  This would just be a special case that would make it easier to avoid the trap of failing to anticipate all the headers that the client might throw into the request.

As you point out in the comments, the AllowedHeaders are unbounded.  So you are already doing this after you check to make sure that the Requested headers are a subset of the Allowed headers.  This would just short-circuit that step by treating this as a special case which would accept any requested headers.  Of course, it could also be extended to a general capability that treats the AllowedHeaders as regexps and actually checks them.  But that is more work and, frankly, greater chance for a mistake or misunderstanding.

I'm not sure if this is a good idea or not.  That's why I'm asking.

**Comment by rs** [comment]
That's definitely an interesting idea. I'm not sure to see the use-case though. That would mean your app doesn't know all the headers that it support?

**Comment by xogeny** [comment]
##### Background (skip of you already know about HATEOAS)

Well, I'm currently building a (soon to be open source) Go library for creating hypermedia APIs.  For me, one of the big wins with hypermedia APIs is the fact that if you follow the HATEOAS approach, your client can be quite generic.  The key is to avoid having to explicitly tell your client about the "routes" or "endpoints" and make the API discoverable.  With my library, you build a bunch of resources and it generates a discoverable API that can then be "picked up" by any client simply by leveraging links/relations (just like a browser).  This is the main idea behind HAL, Siren, etc.  In such cases, your client simply follows links, e.g.,

```
api.follow('ea:orders')
         .withTemplateParameters({ id: 13 })
         .get(callback);  // NO URL BUILDING!
```

##### The Point

The issue I've run into is that there are a couple of Javascript client libraries out there that support different hypermedia formats (HAL, Siren, etc).  When I started using `traverson` (just one example), I noticed it injected its own header.  So I had to include that explicitly when using `cors`.

I guess the bottom line is that once you create a HATEOAS service, the client and the server (by design) really don't need to know much about each other because the formats for the data and the relations are standardized.  So I'm **not** writing the code that makes the requests myself and, as such, I don't necessarily know on the server side what headers might show up.  So I'd just like to make that robust by making it tolerant to add headers.

It isn't that big a deal honestly.  It is just a snag I ran into along the way.  The current work around is that I'd have to try my server with a bunch of different client libraries to make sure they don't add any unusual headers (like `traverson` does).  It's just annoying when you are trying to make an API where the client doesn't need have a bunch of hardwired information about the server in it...except for the allowed headers. ;-)

**Comment by rs** [comment]
Ok it makes much more sense in such context. I'm quite ok to add this feature then. Do you plan on sending a PR?

**Comment by xogeny** [comment]
Closing this since it is addressed by #6.
